lib/linter/linter.js

```diff
@@ -1537,102 +1556,86 @@ class Linter {
         const originalExtname = path.extname(filename);
+        const { files } = preprocessResult;
 
-        let blocks;
-
-        try {
-            blocks = preprocess(text, filenameToExpose);
-        } catch (ex) {
-
-            // If the message includes a leading line number, strip it:
-            const message = `Preprocessing error: ${ex.message.replace(/^line \d+:/iu, "").trim()}`;
-
-            debug("%s\n%s", message, ex.stack);
-
-            return [
-                {
-                    ruleId: null,
-                    fatal: true,
-                    severity: 2,
-                    message,
-                    line: ex.lineNumber,
-                    column: ex.column,
-                    nodeType: null
-                }
-            ];
-        }
-
-        const messageLists = blocks.map((block, i) => {
-            debug("A code block was found: %o", block.filename || "(unnamed)");
+        const messageLists = files.map(block => {
+            debug("A code block was found: %o", block.path || "(unnamed)");
 
             // Keep the legacy behavior.
             if (typeof block === "string") {
                 return this._verifyWithFlatConfigArrayAndWithoutProcessors(block, config, options);
             }
 
-            const blockText = block.text;
-            const blockName = path.join(filename, `${i}_${block.filename}`);
-
             // Skip this block if filtered.
-            if (!filterCodeBlock(blockName, blockText)) {
+            if (!filterCodeBlock(block.path, block.body)) {
                 debug("This code block was skipped.");
                 return [];
             }
 
             // Resolve configuration again if the file content or extension was changed.
-            if (configForRecursive && (text !== blockText || path.extname(blockName) !== originalExtname)) {
+            if (configForRecursive && (text !== block.body || path.extname(block.path) !== originalExtname)) {
                 debug("Resolving configuration again because the file content or extension was changed.");
                 return this._verifyWithFlatConfigArray(
-                    blockText,
+                    block.body,
                     configForRecursive,
-                    { ...options, filename: blockName, physicalFilename }
+                    { ...options, filename: block.path, physicalFilename: block.physicalPath }
                 );
             }
 
+            slots.lastSourceCode = null;
+
             // Does lint.
-            return this._verifyWithFlatConfigArrayAndWithoutProcessors(
-                blockText,
+            return this.#flatVerifyWithoutProcessors(
+                block,
                 config,
-                { ...options, filename: blockName, physicalFilename }
+                { ...options, filename: block.path, physicalFilename: block.physicalPath }
             );
         });
 
         return postprocess(messageLists, filenameToExpose);
```

Comment:
```suggestion
        return processorService.postprocessSync(file, messageLists, {
            processor: {
                preprocess,
                postprocess
            }
        });
```

I'm guessing we'd like to use ProcessorService here as well.